File: interpreter.py
import sys

TOKENS = [
    "PUSH",         # >  0
    "POP",          # <  1
    "ADD",          # +  2
    "SUB",          # -  3
    "PRINT_CHAR",   # @  4
    "PRINT_NUM",    # %  5
    "INPUT",        # #  6
    "INVERT",       # $  7
    "LOOP_START",   # :  8
    "LOOP_END",     # ;  9
    "COMPARE",      # ?  10
    "DUPLICATE",    # =  11
    "END",          # !  12
    "NUMBER"
]

stack = []

class Lexer:
    def __init__(self, line):
        self.line = line

    def tokenize(self):
        tokenized = []
        i = 0
        l = len(self.line)
        while i < l:
            if self.line[i] == '>':
                tokenized.append(TOKENS[0])
                i += 1
            elif self.line[i] == '<':
                tokenized.append(TOKENS[1])
                i += 1
            elif self.line[i] == '+':
                tokenized.append(TOKENS[2])
                i += 1
            elif self.line[i] == '-':
                tokenized.append(TOKENS[3])
                i += 1
            elif self.line[i] == '@':
                tokenized.append(TOKENS[4])
                i += 1
            elif self.line[i] == '%':
                tokenized.append(TOKENS[5])
                i += 1
            elif self.line[i] == '#':
                tokenized.append(TOKENS[6])
                i += 1
            elif self.line[i] == '$':
                tokenized.append(TOKENS[7])
                i += 1
            elif self.line[i] == ':':
                tokenized.append(TOKENS[8])
                i += 1
            elif self.line[i] == ';':
                tokenized.append(TOKENS[9])
                i += 1
            elif self.line[i] == '?':
                tokenized.append(TOKENS[10])
                i += 1
            elif self.line[i] == '=':
                tokenized.append(TOKENS[11])
                i += 1
            elif self.line[i] == '!':
                tokenized.append(TOKENS[12])
                i += 1
            elif self.line[i].isdigit():
                j = i
                while j < l and self.line[j].isdigit():
                    j += 1
                tokenized.append({TOKENS[-1]: int(self.line[i:j])})
                i = j
            elif self.line[i].isspace():
                i += 1
            else:
                raise Exception(f"Invalid instruction: {self.line[i]}")
        return tokenized

class Parser:
    def __init__(self, token_list):
        self.token_list = token_list

    def parse(self):
        i = 0
        while i < len(self.token_list):
            if self.token_list[i] == TOKENS[0]:
                try:
                    stack.append(self.token_list[i+1][TOKENS[-1]])
                    i+=2
                except:
                    raise Exception(f"Invalid token: {self.token_list[i+1]} is not a number")
            elif self.token_list[i] == TOKENS[1]:
                stack.pop()
                i += 1
            elif self.token_list[i] == TOKENS[2]:
                a = stack.pop()
                b = stack.pop()
                c = a+b
                stack.append(c)
                i+=1
            elif self.token_list[i] == TOKENS[3]:
                a = stack.pop()
                b = stack.pop()
                c = a-b
                stack.append(c)
                i+=1
            elif self.token_list[i] == TOKENS[4]:
                char = stack.pop()
                print(chr(char),end='')
                i+=1
            elif self.token_list[i] == TOKENS[5]:
                print(stack.pop(),end='')
                i+=1
            elif self.token_list[i] == TOKENS[6]:
                user_input = input()
                if user_input:
                    stack.append(ord(user_input[0]))
                i+=1
            elif self.token_list[i] == TOKENS[7]:
                stack.reverse()
                i += 1
            elif self.token_list[i] == TOKENS[8]:
                j = i + 1
                depth = 1
                loop_inst = []
                while j < len(self.token_list) and depth > 0:
                    if self.token_list[j] == TOKENS[8]:
                        depth += 1
                    elif self.token_list[j] == TOKENS[9]:
                        depth -= 1

                    if depth > 0:
                        loop_inst.append(self.token_list[j])
                    j += 1
                
                if depth != 0:
                    raise Exception("Unterminated loop")
                
                while stack and stack[-1] != 0:
                    Parser(loop_inst).parse()
                i = j

            elif self.token_list[i] == TOKENS[10]:
                if len(stack) < 2:
                    raise Exception("COMPARE requires at least two arguments")
                
                a = stack[-1]
                b = stack[-2]
                if a == b:
                    i += 1
                else:
                    if self.token_list[i+1] == TOKENS[8]:
                        j = i+2
                        depth = 1
                        while j < len(self.token_list) and depth > 0:
                            if self.token_list[j] == TOKENS[8]:
                                depth += 1
                            elif self.token_list[j] == TOKENS[9]:
                                depth -= 1

                            j += 1

                        i = j
                    else:
                        i += 2
                continue
            elif self.token_list[i] == TOKENS[11]:
                if stack:
                    stack.append(stack[-1])
                else:
                    raise Exception("DUPLICATE on empty stack")
                i += 1
            elif self.token_list[i] == TOKENS[12]:
                sys.exit()
            else:
                raise Exception(f"Invalid token: {self.token_list[i]}")


with open(sys.argv[1]) as code:
    code = code.readlines()
    for line in code:
        lexer = Lexer(line)
        token_list = lexer.tokenize()
        parser = Parser(token_list)

        parser.parse()

File: interpreter.rs
use std::{io::{self, Write}, process::exit, env, fs};


#[derive(Clone)]
enum TokenId {
    Push,
    Pop,
    Add,
    Sub,
    PrintChar,
    PrintNum,
    Input,
    Invert,
    LoopStart,
    LoopEnd,
    Compare,
    Duplicate,
    End,
    Number
}

#[derive(Clone)]
#[allow(dead_code)]
struct Token {
    id: TokenId,
    num: Option<i32>
}

struct Lexer {
    line: String
}

fn make_token(id: TokenId) -> Token {
    Token { id, num: None }
}

impl Lexer {

    fn tokenize(self) -> Vec<Token>{
        let mut tokenized: Vec<Token> = Vec::new();
        let mut i = 0;

        let chars: Vec<char> = self.line.chars().collect();
        while i < chars.len() {
            let c = chars[i];
            match c {
                '>' => tokenized.push(make_token(TokenId::Push)),
                '<' => tokenized.push(make_token(TokenId::Pop)),
                '+' => tokenized.push(make_token(TokenId::Add)),
                '-' => tokenized.push(make_token(TokenId::Sub)),
                '@' => tokenized.push(make_token(TokenId::PrintChar)),
                '%' => tokenized.push(make_token(TokenId::PrintNum)),
                '#' => tokenized.push(make_token(TokenId::Input)),
                '$' => tokenized.push(make_token(TokenId::Invert)),
                ':' => tokenized.push(make_token(TokenId::LoopStart)),
                ';' => tokenized.push(make_token(TokenId::LoopEnd)),
                '?' => tokenized.push(make_token(TokenId::Compare)),
                '=' => tokenized.push(make_token(TokenId::Duplicate)),
                '!' => tokenized.push(make_token(TokenId::End)),
                _ => (),
            }
            
            if c.is_numeric() {
                let mut j: usize = i;
                while j < self.line.len() && self.line.chars().nth(j).unwrap().is_numeric(){
                    j += 1;
                }
                let slice = self.line.get(i..j).expect("slice out of bounds");
                let result: i32 = slice.parse().expect("not a number");
                tokenized.push(Token { id: TokenId::Number, num: Some(result)});

                i = j;
                continue;
            }


            i += 1;
        }

        return tokenized;
    }
}

struct Parser {
    token_list: Vec<Token>
}

impl Parser {
    fn parse(&mut self, stack: &mut Vec<i32>) {
        let tokens: Vec<Token> = self.token_list.clone();
        let mut i = 0;
        while i < tokens.len() {
            match tokens[i].id {
                TokenId::Push => {
                    let next: &Token = &tokens[i+1];
                    stack.push(next.num.unwrap());
                    i += 2;
                },
                TokenId::Pop => {
                    stack.pop();
                    i += 1;
                },
                TokenId::Add => {
                    let a = stack.pop().unwrap();
                    let b = stack.pop().unwrap();
                    let c = a+b;
                    stack.push(c);
                    i += 1;
                },
                TokenId::Sub => {
                    let a = stack.pop().unwrap();
                    let b = stack.pop().unwrap();
                    let c = a-b;
                    stack.push(c);
                    i += 1;
                },
                TokenId::PrintChar => {
                    let ch: i32 = stack.pop().unwrap();
                    print!("{}", ch as u8 as char);
                    io::stdout().flush().unwrap();
                    i += 1;
                },
                TokenId::PrintNum => {
                    let ch: i32 = stack.pop().unwrap();
                    print!("{}", ch);
                    io::stdout().flush().unwrap();
                    i += 1;
                },
                TokenId::Input => {
                    let mut s = String::new();
                    std::io::stdin().read_line(&mut s).unwrap();
                    let ascii = s.as_bytes().get(0).copied();

                    stack.push(ascii.unwrap() as i32);
                    i += 1;
                },
                TokenId::Invert => {
                    stack.reverse();
                    i += 1;
                },
                TokenId::LoopStart => {
                    let mut j = i + 1;
                    let mut depth: i32 = 1;
                    let mut loop_inst: Vec<Token> = Vec::new();

                    while j < tokens.len() && depth > 0 {
                        match tokens[j].id {
                            TokenId::LoopStart => depth += 1,
                            TokenId::LoopEnd => depth -= 1,
                            _ => ()
                        }

                        if depth > 0 {
                            loop_inst.push(tokens[j].clone());
                        }

                        j += 1;
                    }

                    let mut loop_parser: Parser = Parser { token_list: loop_inst };

                    while !stack.is_empty() && stack[stack.len() - 1] != 0{
                        loop_parser.parse(stack);
                    }

                    i = j;
                },
                TokenId::Compare => {
                    if stack.len() < 2 {
                        println!("Compare needs at least two elements");
                        exit(1);
                    }

                    let a = stack[stack.len() - 1];
                    let b = stack[stack.len() - 2];

                    if a == b {
                        i += 1;
                    } else {
                        let next: Option<&Token> = self.token_list.get(i+1);
                        match next.unwrap().id {
                            TokenId::LoopStart => {
                                let mut j = i + 2;
                                let mut depth = 1;
                                while j < self.token_list.len() && depth > 0 {
                                    match self.token_list[j].id {
                                        TokenId::LoopStart => depth += 1,
                                        TokenId::LoopEnd => depth -= 1,
                                        _ => ()
                                    }
                                    j += 1;
                                }
                                i = j;
                            },
                            _ => i += 2
                        }
                    }
                    continue;
                },
                TokenId::Duplicate => {
                    if let Some(&last) = stack.last() {
                        stack.push(last);
                    } else {
                        println!("Duplicate on empty stack");
                        exit(1);
                    }
                    i += 1;
                },
                TokenId::End => exit(0),

                _ => ()
            }
        }
    }
}

fn main() -> io::Result<()> {
    let mut args = env::args();
    let _exe = args.next();
    let filename = match args.next() {
        Some(f) => f,
        None => {
            eprintln!("usage: bolaga_rust <file>");
            return Ok(());
        }
    };

    let code = fs::read_to_string(&filename)?;
    let lexer = Lexer { line: code };
    let tokens = lexer.tokenize();
    let mut parser = Parser { token_list: tokens };
    let mut stack: Vec<i32> = Vec::new();
    parser.parse(&mut stack);
    Ok(())
}
